### ISSUE TITLE
MetricList for moving entire list of metrics onto GPU

### ISSUE THREAD

**Issue by janhenriklambrechts** [PyTorchLightning/pytorch-lightning#5627]
## 🚀 MetricList
MetricList serves the same function for Metrics in PL as `nn.ModuleList` in PyTorch for Modules. 
As such, they take care of flexibly moving all Metrics and their states onto GPU or CPU. 

### Motivation
- As a dynamic inference researcher, I want to be able to test multiple setups for a single model (f.e. executing my model at different depths or widths). 
- For each of these setups, I want to be able to track mostly the same metrics (f.e. Classification Accuracy) and making separate PL Metrics for each one of these metrics is tedious. 
- Moreover, I might want to specify the amount of setups with a parameter which makes constructing them in a `list` format the most elegant solution. 

### Pitch

- In regular PyTorch they introduced `nn.ModuleList` especially for this purpose (f.e. swapping out different FC layers for different setups)

### Alternatives

- Evaluating if any of the internal variables is a list that might contain metrics does not seem like good design to me, but has the advantages that the user will not need to know of the existence of `MetricList`.  Maybe we can add a warning or error to use a MetricList when any errors as per "Additional Context" happens?

### Additional context
- Using regular lists that hold metrics and updating them while training on GPU will now keep metric states on CPU leading to errors, see a `BoringModel` example [here](https://colab.research.google.com/drive/1HvWVVTK8j2Nj52qU4Q4YCyzOm0_aLQF3?usp=sharing), that as expected returns:   `Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!`
-  Happy to work on a PR for this feature, but wanted to know your thoughts.

**Comment by awaelchli** [comment]
Could you add to your description why `nn.ModuleList` doesn't work or reproduce it? Your linked repro script doesn't actually have any metrics or metrics list. Metrics are just modules, so they should work with ModuleList and ModuleDict and if not it's a bug

**Comment by janhenriklambrechts** [comment]
Hi @awaelchli, thanks for checking it out. I noticed I sent the wrong repro script, my bad. 
[This](https://colab.research.google.com/drive/1uNoCoBP3cw09_olfhpvdmGUb-pRX4R2c?usp=sharing) is the correct link, failing with the relatively unhelpful `Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!`. 

I did not know that Metrics are just modules, and adding them to a `nn.ModuleList` indeed works totally fine. Only now I notice that indeed, this is also explained clearly in the docs. I must have read over it. 

That only leaves the question if there is some better information we can give in a stacktrace?

**Comment by awaelchli** [comment]
> That only leaves the question if there is some better information we can give in a stacktrace?

I doubt it, I wouldn't know an elegant way. It seems this would have to be addressed by pytorch directly. 
I can only assume the reason why pytorch does not deal with lists directly is because it can be expensive to recursively traverse the object `__dict__` and it could also have cycles. Just speculation, I did not find much info about that.

**Comment by janhenriklambrechts** [comment]
I see, thank you so much for your fast help. Much obliged! I will close this issue then.

**Comment by awaelchli** [comment]
thanks for reporting a potential issue
